designdocs/provider_permissions.md

```diff
@@ -0,0 +1,207 @@
+
+- ***RESTRICTED***: Where a user grants a provider to publish data on their behalf for specific schema(s) only. This is the default state of a provider on MRC, where a provider has to explicitly provide a list of schema(s) for which they are allowed to publish data on behalf of the user.
+
+An example of permission data structure is as follows:
+
+```rust
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum PermissionType {
+    Publisher,
+    Restricted,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Permission {
+    pub permission_type: PermissionType,
+    pub tos_hash: Vec<u8>,
+    pub expiry_time: u64,
+}
+```
+
+### Grants
+
+Grants enable delegators as well as providers to restrict one another from publishing data on specific schema(s). For version 1 of this implementation, the following options are available:
+
+- ***PUBLISH***: Where a delegator grants a provider to publish data on their behalf for specific schema(s) only. This is the default state of a provider on MRC, where a provider has to explicitly provide a list of schema(s) for which they are allowed to publish data on behalf of the delegator. This also enables a delegator to opt in to publish their data.
+
+- ***BLOCK***: When a delegator or provider want to restrict publication of data on specific schema(s). This is default state of any schemas, not authorized by delegator or provider as part of schema grants request.
+
+An example of grant data structure is as follows:
+
+```rust
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum GrantType {
+    Publish,
+    Block,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Grant {
+    pub grant_type: GrantType,
+    pub tos_hash: Vec<u8>,
+    pub expiry_time: u64,
+}
+```
+
+### API (Extrinsic)
+
+- ***delegator_msa***: The MSA of a delegator/user.
+- ***provider_msa***: The MSA of the provider/app.
+- ***Permission***: The generic option for any user. Default mode of operation is ***Restricted*** for any provider unless explicitly added as ***Publisher***.
+- ***Grant***: The user level action/result. "A user grants a permission to a provider".
+- ***ToS***: The hash of terms of service between a delegator and provider.
+- ***expiry***: The expiry time of a permission/grant.
+- ***schema_id***: The unique identifier of a registered schema on MRC.
+
+### add_schema_grants()
+
+An extrinsic to allow a provider to request publish write to list of schemas. Rendering them **Restricted** status.
+
+- Parameters:
+    1. **provider_msa**: The MSA of the provider/app.
+    2. **delegator_msa**: The MSA of a user.
+    3. **schemas**: The list of schemas for which the provider wants to grant publish write typically ```Vec<SchemaId>```.
+    4. **tos_hash**: The hash of terms of service between a delegator and provider.
+
```

Comment:
provider msa can be extract from origin but these are just placeholders of what api will require from design perspective